Refactor: handle path params

Gayun00 opened the pull request:
### Description
#### **1. Router class 싱글턴 패턴 적용**
##### Problem
- routes와 router관련 로직이 각 컴포넌트에 흩어져 선언되어 있던 문제
- Router에서 관련로직을 관리하고 Router class의 인스턴스를 컴포넌트에서 생성해 사용할 경우 재귀호출로 인한 콜스택 에러 발생

##### Solution
- Router class에 Singleton 패턴 적용
- Router 클래스 내부에서 routes와 관련 로직 관리
관련 이슈: #29 

<br>

#### 2. **모든 path parameter에 대응한 라우트 기능 (handleRoutePath)**
##### Problem
- path parameter의 갯수가 많아질 경우 대응할 수 없음
- 동적 라우팅 경로의 path parameter의 갯수가 많아질 경우 대응할 수 없음
- path parameter의 이름 (pathname)이 달라지는 경우 대응할 수 없음
- path paramter를 받아오는 로직을 컴포넌트에서 호출해 사용하기에 간편하고 직관적이지 않음

##### Solution
- handleRoutePath 메서드: routes배열을 돌면서 인자로 받은 path와 동일한 path가 있다면 반환하고,
  동적 라우팅 주소가 있다면 각 변수에 대응하는 path parameter 값을 찾아 변수:파라미터 값의 객체 형태로 반환
- React-router의 useParams() 메서드처럼 원하는 path parameter의 값을 받아 사용할 수 있도록 리팩토링
예) `const {id} = getParams()`
  - path parameter의 갯수가 많거나 이름이 달라졌을 때, 동적 라우팅 변수가 존재할 때 대응

<br>

### Question
- Router class
  1) Router 클래스 내부에 관련 로직들(예를 들어, app.js내 routes를 선언하지 않고 Router 클래스내에 선언)을 모아놓은 방식이 적절한 방식일지
  2) 이 문제를 해결하기 위해 싱글턴 패턴을 적용한 것이 괜찮은 방법일지 궁금합니다.


- handleRoutePath 메서드 (가장 고민 중...)
  - 내부 코드 자체가 많이 가독성이 떨어지고 깔끔한 방법이 아닌 것 같다는 생각이 드는데, 
  어떤 방법으로 개선해보면 좋을까요? 다른 더 좋은 방향이 있을까요?
  - handleRenderPage와 getPathParams메서드가 각각 handleRoutePath메서드를 공통으로 사용해 라우트 경로에 맞는 페이지 혹은 path parameter를 반환하도록 했는데, 함수가 잘 분리된 것 같지 않아서 더 좋은 방법이 있을지 궁금합니다.

<br>

### Note
- DetailPage, PostTitle 컴포넌트:
 각각의 Post 컴포넌트를 클릭했을 때  해당하는 path parameter인 id값을 받아와 사용할 수 있음을 확인하기 위해 
 임시로 화면에 id값이 표시되도록 간단히 작업해두었습니다.

Gayun00 committed:
Chore: add jest

Gayun00 committed:
Feat: add handlePathParams, getParams

Gayun00 committed:
Feat: add getPathParams

- add getPathParams to refactor dynamic routing logic
- getPathParams can be used to get path parameter of dynamic routed current path

Gayun00 committed:
Correct: add _prefix to private method

Gayun00 committed:
Feat: add handleRenderingPage

- add handleDynamicRoute() to use in getPathParams, handleRenderingPage

Gayun00 committed:
Correct: index.test.js

Gayun00 committed:
Refactor: add handlling static route path

- correct handleDynamicRoute to getMatchedRoute
- getMatchedRoute contains handling static route path

Gayun00 committed:
Refactor: render page

- add attachTo method to Base class
- correct detail page route to /post
- refactor outes to array, use handleRenderingPage to handle all pathname

Gayun00 committed:
Refactor: add singleton pattern to Router class

Related to: #29

Gayun00 committed:
Refactor: dynamic route

- add getParam method to Router class
- display post id in post detail page

close: #29

Gayun00 committed:
Refactor: handlePathParams to handle multiple dynamic route variable

- update test
- refactor func

in progress:
- need to test iterating routes to get matched route
- need to refactor getPathParams, getMatchedRoute
- need to correct func name

Gayun00 committed:
Refactor: add for...of to _checkDynamicRouteVar

- fix issue: return matched route when it is not fully matched route path
- fix issue: not returning value

Gayun00 committed:
Refactor: getPathParams, handleRenderingPage

- correct handleRenderingPage to handleRenderPage
- delete _getMatchedRoute
- correct _checkDynamicRouteVar to handleRoutePath
- refactor _handleRoutePath to return matched path parameter as an object
- update test

BuildTheTruth commented on src/pages/Detail/index.js:
```diff
@@ -1,7 +1,28 @@
+  getParam() {
+    const { id } = Router.getInstance().getParam();
+    this.id = id;
+    this.render(id);
```
해당 컴포넌트 내부에 Id props가 존재하는데 굳이 id를 render함수에 넘겨줄 필요가 없지 않을까 싶습니다.

Gayun00 replied:
맞아요.. 코드를 변경하는 과정에서 불필요한 부분이 남은 것 같습니다.
해당 코드 정리하겠습니다 :)

BuildTheTruth commented on src/utils/handleParams/index.js:
```diff
@@ -0,0 +1,53 @@
+  return route.page;
+};
+
+export const _handleRoutePath = (routes, path) => {
```
_ prefix를 붙인 이유는 무엇인가요? 혹시 private 함수를 표현하고 싶다면 export를 안하면 되지 않을까요?

Gayun00 replied:
index.test.js에서 함수들을 Import해 테스트하기 위해서 export해주었는데,
그러다보니 외부에서 사용하지 않는 함수임을 나타내기 위해 _ prefix를 붙였습니다.
음... 더 좋은 테스트 방식이 있을까요? 🤔

BuildTheTruth replied:
음 그렇군요. 테스트를 위해서 어쩔 수 없었네요. 근데 저라면 그냥 외부로 나가는 함수 단위로 테스팅을 했을 것 같습니다. 그리고 사실 해당 함수들이 util이 아니라 Router 내부 함수로 선언되어야 되지않을까라는 생각이 조금 드네요.

BuildTheTruth commented on src/utils/handleParams/index.js:
```diff
@@ -0,0 +1,53 @@
+export const _handleRoutePath = (routes, path) => {
+  for (let route of routes) {
+    if (route.path.includes(":")) {
+      if (route.path.split("/").length === path.split("/").length) {
```
함수 depth가 깊은 것 같군요. 최대한 if문을 제거하려고 노력해보고 아니면 함수로 분리하는 것도 하나의 방법입니다. if 조건문도 함수로 빼면 조금 더 가독성이 좋아지겠죠?

Gayun00 replied:
넵 함수 뎁스를 줄이는 방향으로 개선해보겠습니다 🙃

BuildTheTruth commented on src/utils/handleParams/index.js:
```diff
@@ -0,0 +1,53 @@
+      if (route.path.split("/").length === path.split("/").length) {
+        const pathRegex = _createPathRegex(route.path);
+        const matchedParamArr = path.match(pathRegex);
+        if (!matchedParamArr) throw new Error("not supported path variable");
```
try-catch문 없이 에러를 던지는 것은 좋지 않다고 생각합니다. 에러 핸들링까지 추가 해주는 건 어떨까요?

Gayun00 replied:
네 반영하겠습니다!

BuildTheTruth commented:
수고많으셨습니다! Router를 싱글톤으로 구현하신건 좋은 생각인 것 같네요. 하나의 앱에 여러 개의 Router가 필요하지는 않을테니깐요.
그런데 제 생각에는 routes는 Router 외부에 선언되는 형태가 더 나은 것 같습니다. Router를 구현하실때 마치 npm module로 나가는 모듈이라고 생각하고 구현하시면 조금 더 나은 형태로 개선하실 수 있으실 것 같네요.

handleRoutePath를 공통으로 사용하여 함수를 분리하는 형태는 나쁘지 않은 것 같습니다. 그런데 반드시 handleRoutePath를 공통화하여 사용하여야 했나라는 의문이 들긴 합니다. 역할을 분명히 구분하고 분리한다면 공통으로 사용하지 않아도 될 것 같은데요?

전반적으로 변수명 네이밍할때 조금더 신경쓰면 좋을 것 같습니다. -Arr는 보통 -s를 표현하고 -Var도 의미없는 suffix인 것 같습니다. 사실 -Obj도 잘 붙이지 않는데 vanillaJS라서 어느정도 의미가 있어보이네요.

Gayun00 committed:
Correct: detail page

- delete id parameter in render method
- call render method in init

Related to: #35

Gayun00 committed:
Correct: add try...catch with calling _handleRoutePath

Related to: #35

Gayun00 commented:
> 수고많으셨습니다! Router를 싱글톤으로 구현하신건 좋은 생각인 것 같네요. 하나의 앱에 여러 개의 Router가 필요하지는 않을테니깐요. 그런데 제 생각에는 routes는 Router 외부에 선언되는 형태가 더 나은 것 같습니다. Router를 구현하실때 마치 npm module로 나가는 모듈이라고 생각하고 구현하시면 조금 더 나은 형태로 개선하실 수 있으실 것 같네요.
> 
> handleRoutePath를 공통으로 사용하여 함수를 분리하는 형태는 나쁘지 않은 것 같습니다. 그런데 반드시 handleRoutePath를 공통화하여 사용하여야 했나라는 의문이 들긴 합니다. 역할을 분명히 구분하고 분리한다면 공통으로 사용하지 않아도 될 것 같은데요?
> 
> 전반적으로 변수명 네이밍할때 조금더 신경쓰면 좋을 것 같습니다. -Arr는 보통 -s를 표현하고 -Var도 의미없는 suffix인 것 같습니다. 사실 -Obj도 잘 붙이지 않는데 vanillaJS라서 어느정도 의미가 있어보이네요.

- Router는 말씀하신대로 모듈로 생각했을 때 확실히 routes는 외부에서 관리하는게 맞겠네요..!
- handleRoutePath 함수를 공통으로 사용하지 않고 각 함수들의 역할을 명확히 해서 로직을 내부에서 작성하는 방식으로 변경해보겠습니다 :) 
- 변수 네이밍 수정하겠습니다! 

**궁금한 점**
다만 -Var의 경우 조금 고민이 되는 부분이 있는데요,
getDynamicRoutingVar의 경우 `/post/:id`와 같이 동적 라우팅 경로에서 `:id`처럼 동적 라우팅 변수만 가져오는 함수인데, params, path가 아닌 해당 변수만 다루는 함수라는 것을 나타내기 위해서 var를 사용했습니다. 이 경우에도 -var를 사용하는 대신 다른 네이밍이 더 좋을까요? 🧐

Gayun00 committed:
Correct: pathParamArr to pathParams

Related to: #35

BuildTheTruth commented:
말씀주신 `:id`는 정확히는 `Path Variables` 입니다. 줄임말로 -Var라고 붙이는건 단순히 변수라는 suffix를 붙인 느낌이라서 지적하였습니다. 의미 명확히하고 줄임말을 최대한 배제하는 방식으로 가는 것을 추천드립니다.
